lib/src/default_index/mod.rs

```diff
@@ -1227,4 +1229,91 @@ mod tests {
+        let id_4 = CommitId::from_hex("444444");
+        let id_5 = CommitId::from_hex("555555");
+        index.add_commit_data(id_0.clone(), new_change_id(), &[]);
+        index.add_commit_data(id_1.clone(), new_change_id(), &[id_0.clone()]);
+        index.add_commit_data(id_2.clone(), new_change_id(), &[id_0.clone()]);
+        index.add_commit_data(id_3.clone(), new_change_id(), &[id_2.clone()]);
+        index.add_commit_data(id_4.clone(), new_change_id(), &[id_1.clone()]);
+        index.add_commit_data(id_5.clone(), new_change_id(), &[id_4.clone(), id_2.clone()]);
+
+        // Helper function to convert commit IDs to/from index positions and call
+        // `heads_from_range_and_filter`.
+        let heads_range = |roots: &[&CommitId],
+                           heads: &[&CommitId],
+                           filter: &dyn Fn(CommitId) -> bool|
+         -> Vec<CommitId> {
+            let roots = roots
+                .iter()
+                .map(|id| index.as_composite().commit_id_to_pos(id).unwrap())
+                .sorted_by_key(|&pos| Reverse(pos))
+                .collect_vec();
+            let heads = heads
+                .iter()
+                .map(|id| index.as_composite().commit_id_to_pos(id).unwrap())
+                .sorted_by_key(|&pos| Reverse(pos))
+                .collect_vec();
+            index
+                .as_composite()
+                .heads_from_range_and_filter::<Infallible>(roots, heads, |pos| {
+                    Ok(filter(index.as_composite().entry_by_pos(pos).commit_id()))
+                })
+                .unwrap()
+                .into_iter()
+                .map(|pos| index.as_composite().entry_by_pos(pos).commit_id())
+                .collect_vec()
+        };
+
+        // heads(::none())
+        assert!(heads_range(&[], &[], &|_| true).is_empty());
+        // heads(all())
+        assert_eq!(
+            heads_range(&[], &[&id_5, &id_3], &|_| true),
+            vec![id_5.clone(), id_3.clone()]
+        );
+        // heads(~id_5)
+        assert_eq!(
+            heads_range(&[], &[&id_5, &id_3], &|id| id != id_5),
+            vec![id_4.clone(), id_3.clone()]
+        );
+        // heads(id_5..)
+        assert_eq!(
+            heads_range(&[&id_5], &[&id_5, &id_3], &|_| true),
+            vec![id_3.clone()]
+        );
+        // heads(id_5.. ~ id_3)
+        assert!(heads_range(&[&id_5], &[&id_5, &id_3], &|id| id != id_3).is_empty());
+        // heads(id_2..id_4)
+        assert_eq!(
+            heads_range(&[&id_2], &[&id_4], &|_| true),
+            vec![id_4.clone()]
+        );
+        // heads(id_2..id_4 ~ id_4)
+        assert_eq!(
+            heads_range(&[&id_2], &[&id_4], &|id| id != id_4),
+            vec![id_1.clone()]
+        );
+        // heads((id_3 | id_1).. ~ id_5)
+        assert_eq!(
+            heads_range(&[&id_3, &id_1], &[&id_5, &id_3], &|id| id != id_5),
+            vec![id_4.clone()]
+        );
+    }
```

Comment:
nit: can you add some tests having duplicates or duplicates filtered out by predicate? something like: `heads(::(5|4) ~ 5)`, `heads(::(5|5))`.